lib/spack/llnl/util/targets.json

```diff
@@ -0,0 +1,750 @@
+        }
+      }
+    },
+    "steamroller": {
+      "from": "piledriver",
+      "vendor": "AuthenticAMD",
+      "features": [
+        "mmx",
+        "sse",
+        "sse2",
+        "sse4a",
+        "abm",
+        "avx",
+        "aes",
+        "pclmulqdq",
+        "cx16",
+        "ssse3",
+        "sse4_1",
+        "sse4_2",
+        "bmi1",
+        "f16c",
+        "fma",
+        "fsgsbase"
+      ],
+      "compilers": {
+        "gcc": {
+          "name": "bdver3",
+          "versions": "4.8:",
+          "flags": "-march={name} -mtune={name}"
+        }
+      }
+    },
+    "excavator": {
+      "from": "steamroller",
+      "vendor": "AuthenticAMD",
+      "features": [
+        "mmx",
+        "sse",
+        "sse2",
+        "sse4a",
+        "abm",
+        "avx",
+        "aes",
+        "pclmulqdq",
+        "cx16",
+        "ssse3",
+        "sse4_1",
+        "sse4_2",
+        "bmi1",
+        "f16c",
+        "fma",
+        "fsgsbase",
+        "bmi2",
+        "avx2",
+        "movbe"
+      ],
+      "compilers": {
+        "gcc": {
+          "name": "bdver4",
+          "versions": "4.9:",
+          "flags": "-march={name} -mtune={name}"
+        }
+      }
+    },
+    "zen": {
+      "from": "excavator",
+      "vendor": "AuthenticAMD",
+      "features": [
+        "bmi1",
+        "bmi2",
+        "f16c",
+        "fma",
+        "fsgsbase",
+        "avx",
+        "avx2",
+        "rdseed",
+        "clzero",
+        "aes",
+        "pclmulqdq",
+        "cx16",
+        "movbe",
+        "mmx",
+        "sse",
+        "sse2",
+        "sse4a",
+        "ssse3",
+        "sse4_1",
+        "sse4_2",
+        "abm",
+        "xsavec",
+        "xsaveopt",
+        "clflushopt",
+        "popcnt"
+      ],
+      "compilers": {
+        "gcc": {
+          "name": "znver1",
+          "versions": "6:",
+          "flags": "-march={name} -mtune={name}"
+        }
+      }
+    },
+    "zen2": {
+      "from": "zen",
+      "vendor": "AuthenticAMD",
+      "features": [
+        "bmi1",
+        "bmi2",
+        "f16c",
+        "fma",
+        "fsgsbase",
+        "avx",
+        "avx2",
+        "rdseed",
+        "clzero",
+        "aes",
+        "pclmulqdq",
+        "cx16",
+        "movbe",
+        "mmx",
+        "sse",
+        "sse2",
+        "sse4a",
+        "ssse3",
+        "sse4_1",
+        "sse4_2",
+        "abm",
+        "xsavec",
+        "xsaveopt",
+        "clflushopt",
+        "popcnt",
+        "clwb"
+      ],
+      "compilers": {
+        "gcc": {
+          "name": "znver2",
+          "versions": "9:",
+          "flags": "-march={name} -mtune={name}"
+        }
+      }
+    },
+    "ppc64le": {
+      "from": null,
+      "vendor": "generic",
+      "features": [],
+      "compilers": {
+        "gcc": {
+          "versions": "4:",
+          "flags": "-mcpu={name} -mtune={name}"
+        }
+      }
+    },
+    "power8": {
```

Comment:
Good catch. That's definitely something to take into account to refactor `archspec` into a library.

> I doubt we'll see the big-endian variants often, but they do exist.

Does anybody have access to any big-endian Power system?  Asking to check how the `/proc/cpuinfo` file looks like there (I suspect it will be the same as for the little-endian) and add unit tests for detection.

Reply:
I added `/proc/cpuinfo` output from one of our (soon to be gone) BG/Q Power7 nodes.  I think the endian determination will have to come from `platform.machine()` -- basically append `le` if `platform.machine()` ends with `le`.

Note: just to confuse this further, it looks like Debian calls `ppc64le` `ppc64el` just to be different.  See [here](https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1818645) for some `uname` output.  I think our library should always say `ppc64le`, even if the OS calls it `ppc64el`.

Reply:
Ouch, we need to get access to a Python interpreter on a Power with a Debian running and check what `platform.machine` returns then. With some luck Python will do the massaging for us and return `ppc64le` but the documentation [is not prescriptive](https://docs.python.org/3.7/library/platform.html#platform.machine) about that.

Reply:
So, I modeled `ppc64` microarchitectures in the JSON file starting from Power7 (let me know if you want to go further back in time). I added a unit test for the BGQ case based on the output you committed, but as I don't have access to machines like that I couldn't check I mocked everything properly. For instance I assume that:
```python
import platform
assert platform.system() == 'Linux'
```
holds for that machine. I'll try to see if I can get a new try account on Piz Daint.

Reply:
Yeah it reports Linux.  We can probably get you LLNL accounts.  I wouldn’t spend too much time on this as I think the code is currently correct, and these machines are going away.

Reply:
Regarding the Debian machine, I’m not sure where I’d find one.  Maybe IBM’s cloud, but  I think it’s just safe to mock something up *in case* we see `platform.machine == ‘ppc64el’`.  I don’t think we need to confirm this one.

Reply:
I think we probably need only a PowerX machine where we can run a container with a Debian image to check what happens there. I'll check if on Davide (power8 system, CINECA) they have Singularity installed and write back here if I have news on that.

Reply:
Ok. That would be great if we can do it but it shouldn’t hold up the merge on this.

Reply:
So, what I see is the following:
```console
$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (AltArch)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (AltArch)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7:server"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

$ uname -a
Linux davidefe02 3.10.0-862.3.2.el7.ppc64le #1 SMP Tue May 22 03:45:03 BST 2018 ppc64le ppc64le ppc64le GNU/Linux

$ singularity shell docker://python
Singularity python_latest.sif:~/singularity> cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

Singularity python_latest.sif:~/singularity> python -c 'import platform; print(platform.machine())'
ppc64le

Singularity python_latest.sif:~/singularity> uname -a
Linux davidefe02 3.10.0-862.3.2.el7.ppc64le #1 SMP Tue May 22 03:45:03 BST 2018 ppc64le GNU/Linux
```
because, of course, the underlying kernel is the same. If Python takes the information from `uname` this test says nothing on what could happen on a native Debian :cry: Do you want to be defensive and implement something just in case, or are we waiting for a bug report before adding code?

Reply:
I think just a simple check to treat `ppc64le` like `ppc64el` should suffice